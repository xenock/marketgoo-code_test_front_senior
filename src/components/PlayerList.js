import React, { useState } from 'react'
import axios from 'axios'
import { Table, TableRow, TableCell, Button, Spinner } from '@marketgoo/ola'
import {
  removePlayerAPI,
  players as storedPlayers
} from '@store/slices/players'
import { useDispatch, useSelector } from 'react-redux'
import { toast } from 'react-toastify'

axios.defaults.baseURL = 'http://localhost:3000'

const PlayerList = _ => {
  const dispatch = useDispatch()
  const players = useSelector(storedPlayers)
  const [synchronizing, setSynchronizing] = useState([])

  const handleDelete = player => {
    setSynchronizing([...synchronizing, player.id])
    dispatch(removePlayerAPI(player))
      .then(response => {
        toast.success(`ðŸ¦„ ${player.name} removed successfully!`)
        setSynchronizing(synchronizing.filter(id => id !== player.id))
      })
      .catch(_ => {
        toast.error('ðŸ¦„ There was an error...')
        setSynchronizing(synchronizing.filter(id => id !== player.id))
      })
  }

  return !players.length ? (
    <Spinner className={null} size='big' />
  ) : (
    <Table caption={null} responsive sticky={false} stiky>
      <thead>
        <TableRow>
          <TableCell header variant='left'>
            Player
          </TableCell>
          <TableCell header variant='left'>
            Team
          </TableCell>
          <TableCell header variant='right'>
            Score
          </TableCell>
          <TableCell header variant='center'>
            Actions
          </TableCell>
        </TableRow>
      </thead>
      <tbody>
        {players.map(player => (
          <TableRow key={player.id || Math.random()}>
            <TableCell variant='multiline'>{player.name}</TableCell>
            <TableCell variant='multiline'>{player.team}</TableCell>
            <TableCell variant='numeric'>{player.score}</TableCell>
            <TableCell variant='action'>
              <Button
                busy={synchronizing.includes(player.id) && 'Synchronizing'}
                variant='primary'
                onClick={() => handleDelete(player)}
              >
                Remove
              </Button>
            </TableCell>
          </TableRow>
        ))}
      </tbody>
    </Table>
  )
}

export default PlayerList
