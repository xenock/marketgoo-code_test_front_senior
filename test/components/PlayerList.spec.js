import React from 'react'
import renderer from 'react-test-renderer'
import playersFixtures from '@fixtures/players.js'
import { useSelector } from 'react-redux'

import PlayerList from '@components/PlayerList.js'

jest.mock('react-redux', () => ({
  useDispatch: () => jest.fn(),
  useSelector: jest.fn(fn => fn())
}))

describe('PlayerList snapshots', () => {
  it('renders a spinner', () => {
    useSelector.mockReturnValue([])
    const tree = renderer.create(<PlayerList />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders a list', () => {
    useSelector.mockReturnValue(playersFixtures)
    const tree = renderer.create(<PlayerList />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
