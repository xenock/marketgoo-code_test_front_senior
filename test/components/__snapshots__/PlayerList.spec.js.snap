// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerList snapshots renders a list 1`] = `
<div
  className="ola_table-container is-responsive"
>
  <table
    className="ola_table"
  >
    <thead>
      <tr
        className="ola_tableRow"
      >
        <th
          className="ola_tableCell is-left"
        >
          Player
        </th>
        <th
          className="ola_tableCell is-left"
        >
          Team
        </th>
        <th
          className="ola_tableCell is-right"
        >
          Score
        </th>
        <th
          className="ola_tableCell is-center"
        >
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="ola_tableRow"
      >
        <td
          className="ola_tableCell is-multiline"
        >
          Peter
        </td>
        <td
          className="ola_tableCell is-multiline"
        >
          Pandas
        </td>
        <td
          className="ola_tableCell is-numeric"
        >
          517
        </td>
        <td
          className="ola_tableCell is-action"
        >
          <button
            className="ola_button is-primary"
            disabled={false}
            onClick={[Function]}
          >
            <span
              className="ola_button-text"
            >
              Remove
            </span>
          </button>
        </td>
      </tr>
      <tr
        className="ola_tableRow"
      >
        <td
          className="ola_tableCell is-multiline"
        >
          Jhon
        </td>
        <td
          className="ola_tableCell is-multiline"
        >
          Cobrakay
        </td>
        <td
          className="ola_tableCell is-numeric"
        >
          490
        </td>
        <td
          className="ola_tableCell is-action"
        >
          <button
            className="ola_button is-primary"
            disabled={false}
            onClick={[Function]}
          >
            <span
              className="ola_button-text"
            >
              Remove
            </span>
          </button>
        </td>
      </tr>
      <tr
        className="ola_tableRow"
      >
        <td
          className="ola_tableCell is-multiline"
        >
          Tommy
        </td>
        <td
          className="ola_tableCell is-multiline"
        >
          Space
        </td>
        <td
          className="ola_tableCell is-numeric"
        >
          716
        </td>
        <td
          className="ola_tableCell is-action"
        >
          <button
            className="ola_button is-primary"
            disabled={false}
            onClick={[Function]}
          >
            <span
              className="ola_button-text"
            >
              Remove
            </span>
          </button>
        </td>
      </tr>
      <tr
        className="ola_tableRow"
      >
        <td
          className="ola_tableCell is-multiline"
        >
          Blair Downs
        </td>
        <td
          className="ola_tableCell is-multiline"
        >
          Rogan Rasmussen
        </td>
        <td
          className="ola_tableCell is-numeric"
        >
          116
        </td>
        <td
          className="ola_tableCell is-action"
        >
          <button
            className="ola_button is-primary"
            disabled={false}
            onClick={[Function]}
          >
            <span
              className="ola_button-text"
            >
              Remove
            </span>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`PlayerList snapshots renders a spinner 1`] = `
<svg
  className="ola_spinner is-big"
  height={140}
  viewBox="0 0 140 140"
  width={140}
>
  <circle
    cx={70}
    cy={70}
    r={62}
  />
</svg>
`;
